pkg/cmd/cli/sa/newtoken.go

```diff
@@ -0,0 +1,208 @@
+	options := &NewServiceAccountTokenOptions{
+		Out: out,
+		Err: os.Stderr,
+	}
+
+	newServiceAccountTokenCommand := &cobra.Command{
+		Use:     fmt.Sprintf(newServiceAccountTokenUsage, name),
+		Short:   newServiceAccountTokenShort,
+		Long:    newServiceAccountTokenLong,
+		Example: fmt.Sprintf(newServiceAccountTokenExamples, fullname),
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(args, f, cmd); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+
+			cmdutil.CheckErr(options.Run())
+		},
+	}
+
+	newServiceAccountTokenCommand.Flags().Int64Var(&options.TimeoutSeconds, "timeout", 100, "the maximum time allowed to generate a token, in seconds")
+	return newServiceAccountTokenCommand
+}
+
+func (o *NewServiceAccountTokenOptions) Complete(args []string, f *clientcmd.Factory, cmd *cobra.Command) error {
+	if len(args) != 1 {
+		return fmt.Errorf("expected one service account name as an argument, got %q", args)
+	}
+
+	o.SAName = args[0]
+
+	client, err := f.Client()
+	if err != nil {
+		return err
+	}
+
+	o.Namespace = cmdutil.GetFlagString(cmd, "namespace")
+	if len(o.Namespace) == 0 {
+		namespace, _, err := f.DefaultNamespace()
+		if err != nil {
+			return fmt.Errorf("could not retrieve default namespace: %v", err)
+		}
+		o.Namespace = namespace
+	}
+
+	o.SAClient = client.ServiceAccounts(o.Namespace)
+	o.SecretsClient = client.Secrets(o.Namespace)
+	return nil
+}
+
+// Run creates a new token secret, waits for the service account token controller to fulfill it, then adds the token to the service account
+func (o *NewServiceAccountTokenOptions) Run() error {
+	serviceAccount, err := o.SAClient.Get(o.SAName)
+	if err != nil {
+		return fmt.Errorf("could not locate service account %q: %v", o.SAName, err)
+	}
+
+	tokenSecret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: osautil.GetTokenSecretNamePrefix(serviceAccount),
+			Namespace:    serviceAccount.Namespace,
+			Annotations: map[string]string{
+				api.ServiceAccountNameKey: serviceAccount.Name,
+			},
+		},
+		Type: api.SecretTypeServiceAccountToken,
+		Data: map[string][]byte{},
+	}
+
+	persistedToken, err := o.SecretsClient.Create(tokenSecret)
+	if err != nil {
+		return err
+	}
+
+	// we need to wait for the service account token controller to make the new token valid
+	options := api.ListOptions{
+		FieldSelector:   fields.SelectorFromSet(fields.Set(map[string]string{"metadata.name": persistedToken.Name})),
+		Watch:           true,
+		ResourceVersion: persistedToken.ResourceVersion,
+		TimeoutSeconds:  &o.TimeoutSeconds,
+	}
+	watcher, err := o.SecretsClient.Watch(options)
+	if err != nil {
+		return fmt.Errorf("could not begin watch for token: %v", err)
+	}
+	defer watcher.Stop()
+
+	timeoutChannel := time.After(time.Duration(o.TimeoutSeconds) * time.Second)
+	watchChannel := watcher.ResultChan()
+	for {
```

Comment:
bad abstraction.  for loop on waitForToken looks wrong.  The waitForToken shouldn't need to be called multiples times.  It should abstract that away from the caller.